cpp/inc/bond/protocol/detail/protobuf_utils.h

```diff
@@ -0,0 +1,299 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+
+#include <bond/core/config.h>
+#include <bond/core/exception.h>
+#include <bond/core/bond_const_enum.h>
+
+#include <boost/static_assert.hpp>
+
+#include <cstring>
+#include <type_traits>
+
+
+namespace bond
+{
+namespace detail
+{
+namespace proto
+{
+    enum class WireType : uint8_t
+    {
+        VarInt          = 0,
+        Fixed64         = 1,
+        LengthDelimited = 2,
+        Fixed32         = 5
+    };
+
+    enum class Encoding : uint8_t
+    {
+        Fixed,
+        ZigZag
+    };
+
+    enum class Packing : uint8_t
+    {
+        False
+    };
+
+    template <typename T>
+    inline BOND_CONSTEXPR T Unavailable()
+    {
+        BOOST_STATIC_ASSERT(std::is_same<T, Encoding>::value
+                            || std::is_same<T, Packing>::value
+                            || std::is_same<T, WireType>::value);
+
+        return static_cast<T>(0xF);
+    }
+
+    BOND_NORETURN inline void NotSupportedException(const char* name)
+    {
+        BOND_THROW(CoreException, name << " is not supported.");
+    }
+
+    BOND_NORETURN inline void ZigZagEncodingException()
+    {
+        BOND_THROW(CoreException, "Unsigned integers cannot have ZigZag encoding.");
+    }
+
+    inline WireType GetWireType(BondDataType type, Encoding encoding)
+    {
+        switch (type)
+        {
+        case BT_BOOL:
+            return WireType::VarInt;
+
+        case BT_FLOAT:
+            return WireType::Fixed32;
+
+        case BT_DOUBLE:
+            return WireType::Fixed64;
+
+        case BT_UINT8:
+        case BT_UINT16:
+        case BT_UINT32:
+        case BT_UINT64:
+            switch (encoding)
+            {
+            case Encoding::Fixed:
+                return type == BT_UINT64 ? WireType::Fixed64 : WireType::Fixed32;
+
+            case Encoding::ZigZag:
+                ZigZagEncodingException();
+                break;
+
+            default:
+                return WireType::VarInt;
+            }
+            break;
+
+        case BT_INT8:
+        case BT_INT16:
+        case BT_INT32:
+        case BT_INT64:
+            switch (encoding)
+            {
+            case Encoding::Fixed:
+                return type == BT_INT64 ? WireType::Fixed64 : WireType::Fixed32;
+
+            case Encoding::ZigZag:
+            default:
+                return WireType::VarInt;
+            }
+            break;
+
+        case BT_STRUCT:
+        case BT_STRING:
+            return WireType::LengthDelimited;
+
+        case BT_WSTRING:
+            NotSupportedException("Wide string");
```

Comment:
Any thoughts on handling `WSTRING` in the same path as `STRING`? Unless you forsee a good way for protobufs to represent wstrings without losing the encoding, I think we should just do that and add "wstrings will be mapped to strings when writing to protobufs" to the docs.